sdk/src/service.rs

```diff
@@ -291,3 +302,145 @@ where
+        // the index of when the plugin was inserted.
+        pub(super) aws_auth_authenticate_inserted_at: Option<usize>,
+        pub(super) aws_auth_authorize_inserted_at: Option<usize>,
+
+        pub(super) plugin_count: usize,
+    }
+
+    impl<Plugin> Builder<Plugin> {
+        /// Apply a new plugin after the ones that have already been registered.
+        pub fn plugin<NewPlugin>(
+            self,
+            plugin: NewPlugin,
+        ) -> Builder<PluginStack<NewPlugin, Plugin>> {
+            Builder {
+                plugin_pipeline: self.plugin_pipeline.push(plugin),
+                aws_auth_authenticate_inserted_at: self.aws_auth_authenticate_inserted_at,
+                aws_auth_authorize_inserted_at: self.aws_auth_authorize_inserted_at,
+                plugin_count: self.plugin_count + 1,
+            }
+        }
+
+        // TODO: The types are obviously wrong but we can't depend on things here, so we just apply
+        // the `IdentityPlugin`.
+        // If building the plugin is fallible, this would return a `Result`.
+        pub fn aws_auth_authenticate(
+            self,
+            _authenticator: String,
+        ) -> Builder<PluginStack<IdentityPlugin, Plugin>> {
+            Builder {
+                plugin_pipeline: self.plugin_pipeline.push(IdentityPlugin),
+                aws_auth_authenticate_inserted_at: Some(self.plugin_count),
+                aws_auth_authorize_inserted_at: self.aws_auth_authorize_inserted_at,
+                plugin_count: self.plugin_count + 1,
+            }
+        }
+
+        pub fn aws_auth_authorize(
+            self,
+            _authorizer: String,
+        ) -> Builder<PluginStack<IdentityPlugin, Plugin>> {
+            Builder {
+                plugin_pipeline: self.plugin_pipeline.push(IdentityPlugin),
+                aws_auth_authenticate_inserted_at: self.aws_auth_authenticate_inserted_at,
+                aws_auth_authorize_inserted_at: Some(self.plugin_count),
+                plugin_count: self.plugin_count + 1,
+            }
+        }
+
+        /// Build the service configuration.
+        ///
+        // This would be infallible if there's nothing in the Smithy model that requires
+        // configuration at runtime.
+        pub fn build(self) -> Result<super::Config<PluginPipeline<Plugin>>, Error> {
+            let mut config_name_to_setter_methods = std::collections::HashMap::new();
+            if self.aws_auth_authenticate_inserted_at.is_none() {
+                config_name_to_setter_methods.insert(
+                    "AWS Auth authenticate is not configured",
+                    "aws_auth_authenticate",
+                );
+            }
+
+            // Authz is not mandatory!
+            // if self.aws_auth_authorize_inserted_at.is_none() {
+            //     config_name_to_setter_methods
+            //         .insert("AWS Auth authorize is not configured", "aws_auth_authorize");
+            // }
```

Comment:
It seems like, requirements-wise, in the general case we have:

- One change in the Smithy model (e.g. a trait applied to the service shape) requires that N things be configured at runtime.
- Each of these N things maps to application of 1 plugin.
    * Each plugin can be mandatory or not.
- These N plugins can be arranged in a graph of N nodes representing the _dependencies_ between the plugins (e.g. authn must come before authz). The two more likely cases are that the graph is entirely disconnected, or it's a single DAG, but in general the graph is a collection of DAGs.

With this info we can easily code generate the `Config`'s builder.

In the _very general_ case, we might consider that changes are related to each other e.g. a service:
* with only trait `T1` applied requires the plugin `P1` to be configured.
* with only trait `T2` applied requires the plugin `P2` to be configured.
* with both traits `T1` and `T2` applied requires no plugin to be configured.

If we aim to tackle these more general cases, I guess in the data model of the code generator we could have a `Change` (horrible name) class that given a `Model`, produces the graph of plugins that need to be applied. So we'd have a hardcoded poset of `Change`s in the code generator such that if `Ci` < `Cj` then `Cj` takes priority and we discard `Ci`. In the end we'd arrive at an unrelated set of changes such that their plugin graphs are disjoint.

We could even lift the hardcoding and have third-parties integrate easily with this code-generated `Config` mechanism if `Change`s had a _common_ interface and a `String` ID. Then `Change`s could live in separate packages and they'd all be loaded in the classpath and define the ordering among themselves. So long as they form a poset it should work.

I'm probably YAGNIing this since we currently ony have this simple auth{n,z} use case, but it'd be fun to know if this is sound.